src/extension/baser.ts

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { Uri, window, workspace } from 'vscode';
+import '../shared/extension';
+import { Store } from './store';
+
+export class Baser {
+    constructor(
+        private readonly distinctLocalNames: Map<string, string>,
+        private readonly store: Pick<Store, 'distinctArtifactNames'>) {
+    }
+
+    private basesArtifactToLocal = new Map<string, string>([]) // <artifactPath, localPath>
+    private updateBases(artifact: string[], local: string[]) {
+        const i = Array.commonLength(artifact.slice().reverse(), local.slice().reverse());
+        this.basesArtifactToLocal.set(
+            artifact.slice(0, -i).join('/'),
+            local.slice(0, -i).join('/'));
+    }
+
+    private validatedPathsArtifactToLocal = new Map<string, string>()
+    private validatedPathsLocalToArtifact = new Map<string, string>()
+    private updateValidatedPaths(artifact: string, local: string) {
+        this.validatedPathsArtifactToLocal.set(artifact, local);
+        this.validatedPathsLocalToArtifact.set(local, artifact);
+    }
+
+    // Other possibilities:
+    // * 1 local -> 2 artifacts (localPath still possible exact match of artPath)
+    // * Actual 0 artifact match
+    // Future:
+    // 2:2 matching
+    // Notes:
+    // If 2 logs have the same path, then likely the same (unless the path is super short)
+    // If 2 logs don't have the same path, they can still potentially be the same match
+    public translateLocalToArtifact(localPath: string): string { // Future: Ret undefined when certain.
+        // Need to refresh on PathMap update.
+        if (!this.validatedPathsLocalToArtifact.has(localPath)) {
+            const {file} = localPath;
+            if ((	// If no workspace, then the open docs (at this moment) become the workspace.
+            // Over-assuming the localPath.name is distinct. There could be 2+ open docs with the same name.
+                !workspace.workspaceFolders?.length ||
+                    this.distinctLocalNames.has(file)
+            ) &&
+                this.store.distinctArtifactNames.has(file)) {
+
+                const artifactPath = this.store.distinctArtifactNames.get(file)!; // Not undefined due to surrounding if.
+                this.updateValidatedPaths(artifactPath, localPath);
+                this.updateBases(artifactPath.split('/'), localPath.split('/'));
```

Comment:
>/ [](start = 75, length = 1)

is this a file path or URI? Would this work on Windows?

Reply:
Theses are URI paths (soon they will be whole URIs) and are platform agnostic.

---
In reply to: [442392608](https://github.com/microsoft/sarif-vscode-extension/pull/282#discussion_r442392608) [](ancestors = 442392608)